---

output: html_document 

---

```{r global_options, include=FALSE}


knitr::opts_chunk$set(
    warning=FALSE,
    message=FALSE
    )


```


## Loading packages

```{r loading_packages}

library(rmarkdown)
library(tidyverse)
library(ggplot2)
library(Rsubread)
library(AnnotationHub)
library(DESeq2)
library(pheatmap)
library(data.table)
library(gridExtra)
library(ggrepel) 

```

## Setting AnnotationHub
### Assign your species of interest


```{r annotationhub_setup}


AnnotationSpecies <- "Homo sapiens"  # Assign your species 
ah <- AnnotationHub(hub=getAnnotationHubOption("URL"))   # Bring annotation DB

```

## Running AnnotationHub

```{r running_annotationhub}


# Filter annotation of interest
ahQuery <- query(ah, c("OrgDb", AnnotationSpecies))      

if (length(ahQuery) == 1) {
    DBName <- names(ahQuery)
} else if (length(ahQuery) > 1) {
               DBName <- names(ahQuery)[1]
} else {
    print("You don't have a valid DB")
    rmarkdown::render() 
} 
AnnoDb <- ah[[DBName]] # Store into an OrgDb object  


# Explore your OrgDb object with following accessors:
# columns(AnnpDb)
# keytypes(AnnoDb)
# keys(AnnoDb, keytype=..)
# select(AnnoDb, keys=.., columns=.., keytype=...)
AnnoKey <- keys(AnnoDb, keytype="ENSEMBL")
# Note: Annotation has to be done with not genome but transcripts 
AnnoDb <- select(AnnoDb, 
                 AnnoKey,
                 keytype="ENSEMBL",
                 columns="SYMBOL")


head(AnnoDb)
```




## Setting featureCounts parameters 

Defined by users 

**Ensembl Hg19 GTF** was downloaded from below:    
https://hgdownload.soe.ucsc.edu/goldenPath/hg19/bigZips/genes

```{r setting_parameters}

Samples <- c(paste0("mock", 1:3), paste0("CoV", 1:3))

NameTail=".sorted.bam"

BAMInputs <- c()   # Path to Input files 

for (i in 1:length(Samples)) {

    BAMInputs[i] <- paste0("hisat2_output/", 
                           Samples[i],
                           NameTail)
}

# "mm10", "mm9", "hg38", or "hg19"
annot.inbuilt="hg38" 

# annotation data such as a data frame or a GTF file
annot.ext="reference_gtf/gencode.v35.primary_assembly.annotation.gtf.gz"

# annotation type: choose from the GTF file 
# e.g. "gene_id", "transcript_id", "gene_name"
GTF.attrType="gene_name"

# number of cores to run featureCounts()
nthreads=8

```

## Running featureCounts 
### The package is available for Unix and R. For more info, visit below:   
http://subread.sourceforge.net/



```{r running_featureCounts}


# Run featureCounts()
FC <- featureCounts(files=BAMInputs,
                    annot.inbuilt=annot.inbuilt,
                    annot.ext=annot.ext,
                    GTF.attrType=GTF.attrType,
                    isGTFAnnotationFile=TRUE,
                    nthreads=nthreads,
                    verbose=TRUE)

# Extract counts and store into a matrix
FC <- FC$counts 

# Extract feature (gene or transcripts) names
Gene=rownames(FC)

# Create a data frame containing feature names and corresponding counts
CountTable <- cbind(Gene, as.data.frame(FC))


# Change column names 
colnames(CountTable) <- c("Gene", Samples)

head(CountTable)
dim(CountTable)
```



## Saving count data as a csv file


```{r saving_countdata}

# Create a directory to save csv files
dir.create("csv")

# Save as a csv file
write.csv(CountTable, "csv/read_count.csv") 


```

## Creating metadata for DE analysis


```{r creating_metadata}

# Define group level
GroupLevel <- c("Mock", "COVID")

# Define contrast for DE analysis
Contrast <- c("Group", "COVID", "Mock")

# Define sample groups
group <- c(rep("Mock", 3), rep("COVID", 3))


# Create metadata
metadata <- data.frame(Sample=factor(Samples, levels=Samples),
                       Group=factor(group, levels=GroupLevel))

# Assign row names
rownames(metadata) <- metadata$Sample


print(metadata)

```



## Plotting library size per sample

```{r library_size}


# Create a library size table
LibSizeTable <- colSums(CountTable[Samples], 
                        na.rm=TRUE) 
# Data Cleaning
LibSize <- data.frame(Read=LibSizeTable) %>%
    rownames_to_column(var="Sample") %>%
    inner_join(metadata[, c("Sample", "Group")],
               by="Sample")


# Creat a bar plot presenting library size of the dataset
LibSizePlot <- 
    ggplot(LibSize,
       aes(x=Sample,
           y=Read,
           fill=Group,
           label=round(Read))) +
           geom_bar(stat="identity", width=0.8) +
           ggtitle("Library Size") +
           ylab("Number of Total Counts") +
           theme_bw() + 
           scale_y_log10() +
           geom_text(vjust=1.5) +
           theme(axis.text.x=element_text(angle=45, 
                                          vjust=0.5))
# Print the plot
print(LibSizePlot)



```

## Plotting distribution of counts per sample



```{r count_distribution}


# Data Cleaning
ctDistribution <- gather(CountTable,
                         "Sample",
                         "Read",
                         -Gene) 

# Create a density plot presenting distribution of counts
CountDistPlot <-
    ggplot(ctDistribution,
       aes(x=Read,
           color=Sample)) + 
           geom_density(aes(y=..count..), alpha=0.5) + 
           theme_bw() +
           ggtitle("Distribution of Counts") + 
           xlab("Number of Counts") +
           ylab("Number of Genes") + 
           scale_x_log10()

# Print the plot       
print(CountDistPlot)




```

## Data cleaning prior to DE analysis 


```{r deseq_datacleaning}


# Build a count matrix without Transcript/Gene
inputMatrix <- CountTable[, Samples]


# Check out the cleaned TPM matrix
head(inputMatrix)


```

## Creating a DESeq object and VST


```{r creating_DESeqObject}




dds <- DESeqDataSetFromMatrix(inputMatrix, 
                              colData=metadata,
                              design=~Group)


vsd <- varianceStabilizingTransformation(dds,
                                         blind=TRUE) 

```

## Estimating size factors


```{r sizefactors}


# Calculate and add size factors to the DEseq object
dds <- estimateSizeFactors(dds)

# Extract and save the size factors as a data frame
sizeFactor <- as.data.frame(round(sizeFactors(dds), 3))

colnames(sizeFactor) <- 'Size_Factor'

sizeFactor <- sizeFactor %>%
    rownames_to_column(var="Sample") %>%
    inner_join(metadata[, 1:ncol(metadata)], by="Sample") 

# Create a plot comparing the size factors by sample
SizeFactorPlot <- 
    ggplot(sizeFactor, aes(x=Sample, 
                       y=Size_Factor, 
                       fill=Group,
                       label=Size_Factor)) +
    geom_bar(stat="identity", width=0.8) +
    theme_bw() + 
    ggtitle("Size Factors") +
    geom_text(vjust=1.5) +
    theme(axis.text.x=element_text(angle=45, 
                                   vjust=0.5)) + 
ylab("Size Factor")


# Print the plot 
print(SizeFactorPlot)


```

## Estimating dispersions and conducting the Wald Test


```{r dispersion_waldtest}

# Calculate and add dispersions to the DEseq object
dds <- estimateDispersions(dds)

# Add wald test to the DESeq object
dds <- nbinomWaldTest(dds)

```

## Sample QC: Principal Component Analysis 


```{r QC_PCA}


# Assigne what to compare
GroupOfInterest <- Contrast[1]


# Create a PCA plot 
QCPCAPlot<- 
    plotPCA(vsd,
        intgroup=GroupOfInterest,
        returnData=FALSE) +
theme_bw() +
ggtitle("PCA")

# Print the plot
print(QCPCAPlot)



```



## Sample QC: Sample Correlation Heatmap


```{r QC_correlation_heatmap}

# Heatmap annotation
colOfinterest <- !colnames(metadata) %in% c("Sample")
HeatmapAnno <- as.data.frame(metadata[, colOfinterest])
rownames(HeatmapAnno) <- Samples
colnames(HeatmapAnno) <- colnames(metadata)[colOfinterest]


# Extract a normalized count matrix
vsdMatrix <- assay(vsd)
corMatrix <- cor(vsdMatrix)
pheatmap(corMatrix,
         annotation=HeatmapAnno,
         main="Sample Correlation Heatmap")
```


## Running DE analysis


```{r DE_analysis}

# Run DESeq 
dds <- DESeq(dds)

# Check result names 
ResNames <- resultsNames(dds)

print(ResNames)
```

## Creating a dispersion plot

```{r dispersion_plot}

plotDispEsts(dds, 
main="Dispersion over Counts")


```

## Setting how to extract fold-change results
### Change variables below

```{r setting_resultcondition}

# Set the threshold of FDR as a variable "alpha" 
alpha=0.1

# Set the coefficients to compare 
Coef <- ResNames[-1]
print(Coef) 

# Set a function to clean result table 
LFCTable_fn <- function(df) {
    df <- df %>% 
        rownames_to_column(var="Gene") %>%
        mutate(FDR=ifelse(padj < 0.1 & !is.na(padj), 
                                   "< 0.1", 
                                   "> 0.1")) 
    return(df)
}

```


## Extracting log2FoldChanges
### You can change alpha depending on your interest of FDR level



```{r DEresult_extraction}

# Extract DE results
# The Contrast variable was defined in the previous chunk

# Extraction with no shrinkage
# alpha: FDR threshold
Res <- results(dds, contrast=Contrast, alpha=alpha)

# Convert the LFC data to a data frame
ResDF <- LFCTable_fn(as.data.frame(Res))

# Save the LFC data 
write.csv(ResDF, "./csv/LFC_noshrinkage.csv")

# Extract shrunken data 
Res_shr <- lfcShrink(dds, coef=Coef, type="apeglm")

```

## Determining what comparison to explore 
### Checkout resList in the previous chunk and save it to a data frame 


```{r LFC_to_dataframe}

# Save data of interest as a data frame for further analysis and clean the data frame 
shRes <- LFCTable_fn(as.data.frame(Res_shr))

# Save the LFC table
write.csv(shRes, "./csv/LFC_shrinkage.csv")
head(shRes)
```

## Exploring distribution of false discovery rate (FDR)

```{r FDR_distribution}

# Create a plot presenting distribution of FDR
FDR_distPlot <- 
    ggplot(shRes,
           aes(x=padj)) + 
geom_density(aes(y=..count..)) + 
theme_bw() +
ggtitle("Distribution of False Discovery Rate (FDR)") + 
xlab("Adjusted P-Value") + 
ylab("Number of Genes") + 
geom_vline(xintercept=alpha, color="red") +
scale_y_log10()

# Print the plot
print(FDR_distPlot)

```


## Exploring distribution of log2FoldChange

### Black: total genes (padj =/= NA)
### Colored: genes above or below FDR=0.1


```{r L2FC_distribution}
# Subset genes whose padj are not NA
shRes_nonNa <- subset(shRes, !is.na(padj))

L2FC_dist <- 
    ggplot(shRes_nonNa,
           aes(x=log2FoldChange)) + 
geom_density(color="black", aes(y=..count..)) + 
geom_density(data=shRes_nonNa,
             aes(x=log2FoldChange,
                 color=FDR,
                 y=..count..)) +
theme_bw() +
ggtitle("Distribution of log2FoldChange Values") + 
ylab("Number of Genes")

print(L2FC_dist)
```


## Exploring mean-difference with an MA plot


```{r MAplot}
# Define a function creating an MA plot
MA_fn <- function(df, tit) {
    ggplot(df, 
           aes(x=baseMean,
               y=log2FoldChange,
               color=FDR)) +
geom_point()+ 
scale_x_log10() + 
theme_bw() + 
scale_color_manual(values=c("blue", "grey")) + 
ggtitle(tit) + 
geom_hline(yintercept=c(-1, 1), 
           color="red", 
           linetype="dashed") + 
ylim(-8, 8)    # has to be adjusted by users
}

# Create MA plots with or without shrinkage
MAplot_noshr <- MA_fn(ResDF, "Mean-Differene without shrinkage")
MAplot_shr <- MA_fn(shRes, "Mean-Difference with shrinkage")


# Print the plots
grid.arrange(MAplot_noshr, MAplot_shr, ncol=2)
```


## Volcano plots with or without gene names

```{r volcano_plot}

volcano.fn <- function(DF, Title, Min, Max) {

ggplot(DF, 
       aes(x=log2FoldChange,
           y= -log10(padj),
           color=FDR,
           label=Label)) + 
geom_point() +
theme_bw() +
scale_color_manual(values=c("blue", "grey")) + 
ggtitle(Title) + 
ylab("-log10(padj)") + 
xlim(Min, Max) + 
geom_vline(xintercept=c(-1, 1), 
           color="red", 
           linetype="dashed")
}




# Assign log odds threshold 
LogOddsCut=5  # has to be set by users

# Add a column indicating high log-odds genes 
shRes <- shRes %>% 
    mutate(Label=ifelse(-log10(padj) > LogOddsCut, 
                                   Gene, 
                                   "")) 
# Display volcano plots 
withoutGenes <- volcano.fn(shRes, "Volcano Plots", -5, 5)
withGenes <- withoutGenes + geom_text_repel(color="black")

grid.arrange(withoutGenes, withGenes, ncol=2)
```


## Exploring expression profiling (FDR < 0.1)


```{r expression_profiling_FDR}

# Determine row numbers whose FDR is below alpha 
RowBelowAlpha <- which(shRes$FDR == "< 0.1")

# Subset normalized counts with FDR below alpha 
GenesBelowAlpha <- assay(vsd)[RowBelowAlpha,]

# Create a heatmap from GenesBelowAlpha 
pheatmap(GenesBelowAlpha,
         annotation=HeatmapAnno,
         scale="row",    # presents z-score
         border_color=NA,
         main="Gene Expression Profiles (FDR < 0.1)")
```


## Exploring expression profiling 
## (FDR < 0.1 & log2FoldChange >= 1)


```{r expression_profiling_FDRandFold}

# Set minimum log2FoldChange of your interest: MinL2F
MinL2F <- 1

# Determine row numbers whose log2FoldChange >= MinL2F
AboveMinL2F <- which(shRes$FDR == "< 0.1" &
                      shRes$log2FoldChange >= MinL2F)

# Subset normalized counts with log2FoldChange above MinL2F 
GenesAboveMinL2F <- assay(vsd)[AboveMinL2F,]
pheatmap(GenesAboveMinL2F,
         annotation=HeatmapAnno,
         scale="row",
         border_color=NA,
         main="Gene Expression Profiles (FDR < 0.1 and log2FoldChange >= 1)")


length(AboveMinL2F)
```

## NA statistics: zero count genes & outlier genes

### When NAs appear in 
#### **log2FoldChange**: zero counts in all samples
#### **padj**: too little information
#### **pval & padj**: at least one replicate was an outlier

```{r NA_statistics}

# Count number of NA genes  
type=c("Zero Counts", "Outliers", "Total NA Genes") 
NAstat <- shRes %>%
    summarize(zero=sum(is.na(log2FoldChange)), 
              outlier=sum(is.na(pvalue) & is.na(padj))) %>%
    mutate(total=sum(zero, outlier)) %>%
    gather(Type, Number) %>%
    mutate(Type=factor(case_when(Type == "zero" ~ type[1], 
                                 Type == "outlier" ~ type[2], 
                                 Type == "total" ~ type[3]), 
                       levels=type))
# Plot number of NA genes 
ggplot(NAstat, aes(x=Type, y=Number, label=Number)) + 
    geom_bar(stat="identity") + 
    theme_bw() +
    geom_text(position=position_dodge(width=1), vjust=-0.4) + 
    ggtitle("Number of NA Genes") + 
    ylab("Number of Genes")

```

## Session info

```{r session_info}

sessionInfo()


```
